### PR TITLE
Holiday Dungeon

### PR DESCRIPTION
**Pull request by UnknownVoid**
Might require some adjustments to this Puzzle to get it working without errors.

Keychest Skeleton 2 Puzzle is missing.

What was added

5 Races for Holiday Math Run (Normal Run)
Modified Dungeons.txt to add Missing Dungeon
Added Script for the Holiday dungeon

What's Left to do?

Accurate Loot for Holiday dungeon (I have run the dungeon a few times, there probally is a better way of this this...)

Ability to Override Weapon types in script for Monsters that can have weapons (would be a nice feature)

Advanced Holiday Math Dungeon

AI for Kobold and Kobold Archers.

### CODE REVIEW

**Comment by exectails** [comment]
The dungeon is invalid, several scripts are missing mob groups. You should use the client db as base.

>I have run the dungeon a few times, there probally is a better way of this this...)

It's not on the Wiki yet?

>Ability to Override Weapon types in script for Monsters that can have weapons

What does this mean?

**Comment by UnknownVoid** [comment]
well "Ability to Override Weapon types in script for Monsters that can have weapons", what I ment by this is a scripter would not try to override weapons for rats or mobs that can not have anything in there left or right hands.

However there should be some way to replace the weapons or even gear defined during the mobs spawn prossess, so we could have the same mob type with a completely different weapon / armor without having to alter the core race define. (we probally won't ever need this for offical creatures but would be nice for Unoffical ones).

**Comment by exectails** [comment]
You have full control over what's happening in the puzzles, it's not the core that does that. Just don't use the convenience functions, but create, customize, and spawn the monsters yourself.

File: system/scripts/ai/hellhound.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	{
+		Do(Wander());
+		Do(Wait(2000, 5000));
+	}
+	
+	protected override IEnumerable Alert()
+	{
+		if(Random() < 80)
+		{
+			if(Random() < 20)
```
Comment (exectails) [inline system/scripts/ai/hellhound.cs]: Why didn't you do this properly, like in Aggro?

File: system/scripts/ai/hellhound.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+		if(Random() < 50)
+		{
+			var rndnum = Random();
+			if(rndnum < 20) // 20%
+			{
+				Do(PrepareSkill(SkillId.Defense));
+				Do(Circle(500, 1000, 5000));
+				Do(CancelSkill());
+			}
+			else if(rndnum < 35) // 30%
```
Comment (exectails) [inline system/scripts/ai/hellhound.cs]: 15%

File: system/scripts/ai/hellhound.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+			{
+				Do(PrepareSkill(SkillId.Smash));
+				Do(Attack(1, 5000));
+				Do(Wait(3000, 8000));
+			}		
+			else if(rndnum < 42) // 40%
```
Comment (exectails) [inline system/scripts/ai/hellhound.cs]: 7%

File: system/scripts/ai/hellhound.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+			{
+				Do(PrepareSkill(SkillId.Firebolt));
+				Do(Wait(1000, 1300));
+				Do(PrepareSkill(SkillId.Firebolt));
+				Do(Wait(1000, 1300));
+				Do(PrepareSkill(SkillId.Firebolt));
+				Do(Wait(1000, 1300));
+				Do(Attack(1, 5000));
+			}
+			else // 40%
```
Comment (exectails) [inline system/scripts/ai/hellhound.cs]: 50%

File: system/scripts/dungeons/dunbarton_math_crimath_dungeon.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+			drops.Add(new DropData(itemId: 51102, chance: 15, amountMin: 3, amountMax: 5)); // Mana Herb
+			drops.Add(new DropData(itemId: 64020, chance: 18, amountMin: 3, amountMax: 3)); // Iron Ore
+			drops.Add(new DropData(itemId: 64021, chance: 13, amountMin: 3, amountMax: 3)); // Copper Ore
+			drops.Add(new DropData(itemId: 64022, chance: 13, amountMin: 3, amountMax: 3)); // Silver Ore
+			drops.Add(new DropData(itemId: 64023, chance: 10, amountMin: 3, amountMax: 3)); // Gold Ore
+			drops.Add(new DropData(itemId: 64024, chance: 10, amountMin: 3, amountMax: 3)); // Mythril Ore
+			drops.Add(new DropData(itemId: 60051, chance: 5, amountMin: 2, amountMax: 2));  // Tough String
+			drops.Add(new DropData(itemId: 92730, chance: 100, amountMin: 1, amountMax: 1)); // Holiday Gift Box
+					
+			if (IsEnabled("TalentRenovationCloseCombat"))
```
Comment (exectails) [inline system/scripts/dungeons/dunbarton_math_crimath_dungeon.cs]: This is not Python, you have to use curly braces to make the entire block conditional =P

File: system/scripts/ai/hellhound.cs (hunk 1 above)
Comment (UnknownVoid) [inline system/scripts/ai/hellhound.cs]: Fixed, I guess i was messing around with the random values too much and deleted one too many lines was mostly testing The AI pattern and comparing it to official.

It still needs some small adjustments to make it just right. :+1:

**Pull request closed without merging**